astropy/utils/tests/test_data.py

```diff
@@ -1879,96 +1970,119 @@ def test_download_parallel_respects_pkgname(temp_cache, valid_urls):
+
+
+def test_zero_remote_timeout(temp_cache, valid_urls):
+    u, c = next(valid_urls)
+    with pytest.raises(urllib.error.URLError):
+        download_file(u, timeout=0)
+    assert not is_url_in_cache(u)
+    with pytest.raises(urllib.error.URLError):
+        # This will trigger the remote data error if it's allowed to touch the internet
+        download_file(TESTURL, timeout=0)
 
 
-# What happens when the lock can't be obtained in a timely manner?
-def test_lock_unavailable_raises_runtimeerror_and_clear_frees_lock(
-        temp_cache, valid_urls):
+def test_clear_download_cache_not_too_aggressive(temp_cache, valid_urls):
     u, c = next(valid_urls)
     download_file(u, cache=True)
-    with _cache_lock("astropy", need_write=True):
-        with conf.set_temp("download_cache_lock_attempts", 0):
-            with pytest.raises(RuntimeError):
-                with _cache_lock("astropy", need_write=True):
-                    pass
-            # Ensure that the cache doesn't accidentally get unlocked!
-            with pytest.raises(RuntimeError):
-                with _cache_lock("astropy", need_write=True):
-                    pass
-            # Trying to do anything should raise an exception
-            with pytest.raises(RuntimeError):
-                is_url_in_cache(u)
-            with pytest.raises(RuntimeError):
-                download_file(u, cache=True, sources=[])
-            with pytest.raises(RuntimeError):
-                download_file(u, cache=True)
-            clear_download_cache()  # breaks lock
-            is_url_in_cache(u)  # False but doesn't raise an exception
-    # Exiting the lock should succeed even though it was broken open
+    dldir = _get_download_cache_loc()
+
+    bad_filename = os.path.join(dldir, "contents")
+    assert is_url_in_cache(u)
+    clear_download_cache(bad_filename)
+    assert is_url_in_cache(u)
+
+
+def test_clear_download_cache_variants(temp_cache, valid_urls):
+    # deletion by contents filename
+    u, c = next(valid_urls)
+    f = download_file(u, cache=True)
+    clear_download_cache(f)
+    assert not is_url_in_cache(u)
+
+    # deletion by url filename
+    u, c = next(valid_urls)
+    f = download_file(u, cache=True)
+    clear_download_cache(os.path.join(os.path.dirname(f), 'url'))
+    assert not is_url_in_cache(u)
+
+    # deletion by hash directory name
+    u, c = next(valid_urls)
+    f = download_file(u, cache=True)
+    clear_download_cache(os.path.dirname(f))
+    assert not is_url_in_cache(u)
+
+    # deletion by directory name with trailing slash
+    u, c = next(valid_urls)
+    f = download_file(u, cache=True)
+    clear_download_cache(os.path.dirname(f)+'/')
+    assert not is_url_in_cache(u)
+
+    # deletion by hash of file contents
+    u, c = next(valid_urls)
+    f = download_file(u, cache=True)
+    h = compute_hash(f)
+    clear_download_cache(h)
+    assert not is_url_in_cache(u)
+
+
+@pytest.mark.remote_data
+def test_ftp_tls_auto(temp_cache):
```

Comment:
This is the only test in `astropy/utils/tests/test_data.py` that tests the correct behaviour of FTP over TLS - which is a non-trivial feature implemented using a custom wrapper of the internals of `urllib`. It was added without a test in #9964 and this is currently the only one we've got. 

What resources does Astropy have as far as servers go? Certain of the `remote-data=astropy` tests hit Astropy servers, but at the moment I believe it is all `http`. Can Astropy provide a TLS-only FTP server even if it serves only one file?

I can mock this but I think we need a test to verify that the claimed feature - FTP over TLS - actually works at all.

Reply:
I guess it is tested indirectly, because all the remote data jobs that rely on IERS would fail if this doesn't work. I doubt Astropy wants to serve a fake FTP TLS just for this test, so if you insist to keep it, then maybe `xfail` for CI? Either using `CI` or `TRAVIS` will work, see https://docs.travis-ci.com/user/environment-variables/#default-environment-variables .

Reply:
In the interest of being able to identify the problem, I think a unit test that tests specifically the ability of `download_file` to do FTP over TLS is useful.

Should I go and add the `xfail` to all the relevant IERS tests as well?

Reply:
Interestingly enough, only this one test fails with Bad IP, so I don't think you have to xfail the other ones.

Reply:
I put this in because faailing IERS tests made me realize that we weren't testing FTP over TLS. Now I think that test is also xfailed on Travis.